packages/typescript-estree/src/parser.ts

```diff
@@ -125,11 +126,28 @@ function parse<T extends TSESTreeOptions = TSESTreeOptions>(
   return ast;
 }
 
+function isESM<T extends TSESTreeOptions = TSESTreeOptions>(
+  options: T | undefined,
+): boolean {
+  const extension = path
+    .extname(options?.filePath ?? '')
+    .toLowerCase() as ts.Extension;
+
+  return extension === ts.Extension.Mjs || extension === ts.Extension.Mts;
+}
+
 function parseWithNodeMapsInternal<T extends TSESTreeOptions = TSESTreeOptions>(
   code: ts.SourceFile | string,
   options: T | undefined,
   shouldPreserveNodeMaps: boolean,
 ): ParseWithNodeMapsResult<T> {
+  /**
+   * Ensure that `mjs`/`mts` files are always treated as ESM
+   */
+  if (typeof code === 'string' && isESM(options)) {
+    code += '\n' + 'export {}';
```

Comment:
We cannot change the code at all.
Doing so changes the resulting AST and its ranges.
This may also cause lint rules to report on code that doesn't exist.

We need to accomplish this without mutating the passed code at all.

Reply:
Ah! I hadn't thought of that, you're right.

I'll think of another way to do it.

Thanks for the review :)